### PR TITLE
LoopOverLength check

### PR DESCRIPTION
**Pull request by pfitzseb**
Fixes https://github.com/julia-vscode/StaticLint.jl/issues/337.

### CODE REVIEW

**Comment by non-Jedi** [comment]
Does `check_loop_iter` get called on comprehensions or only `for` loops?

**Comment by pfitzseb** [comment]
Ah, good call. Generators and for loops with multiple iterators now work:
image

**Comment by fredrikekre** [comment]
Shouldn't it check that the index is used for indexing? There is nothing wrong with having a loop from 1 to the length of something.

**Comment by pfitzseb** [comment]
Yes.
image

**Comment by pfitzseb** [comment]
Also fixed the wording:
image

**Comment by non-Jedi** [comment]
There's a bikeshedding discussion to be <del>had</del>avoided on whether the linter should highlight the for-loop or the indexing location.

Thanks so much for tackling this @pfitzseb. Hopefully this is one of those little changes that nudge the whole ecosystem towards more care for correctness.

File: src/linting/checks.jl [hunk 1]
```diff
@@ -58,8 +61,9 @@ const LintCodeDescriptions = Dict{LintCodes,String}(IncorrectCallArgs => "Possib
     UnassignedKeywordArgument => "Keyword argument not assigned.",
     CannotDefineFuncAlreadyHasValue => "Cannot define function ; it already has a value.",
     DuplicateFuncArgName => "Function argument name not unique.",
-    IncludePathContainsNULL => "Cannot include file, path cotains NULL characters."
-    )
+    IncludePathContainsNULL => "Cannot include file, path cotains NULL characters.",
```
Comment (fredrikekre) [inline src/linting/checks.jl:64]:
```suggestion
    IncludePathContainsNULL => "Cannot include file, path contains NULL characters.",
```

File: src/linting/checks.jl [hunk 2]
```diff
@@ -58,8 +61,9 @@ const LintCodeDescriptions = Dict{LintCodes,String}(IncorrectCallArgs => "Possib
...
+    LoopOverLength => "Indexing into an array with indices obtained from `for i in 1:length(...)` and similar is discouraged. Use `eachindex` instead."
```
Comment (fredrikekre) [inline src/linting/checks.jl:65]: Possible to have a better name since `LoopOverLength` shows in the UI? I don't  have a good suggestion though...

**Review by fredrikekre — APPROVED** [review]
LGTM but don't know this code that well.

File: src/linting/checks.jl (hunk 2 above)
Comment (pfitzseb) [reply src/linting/checks.jl]: Changed it to `ArrayIndexFromLength`.

**Comment by pfitzseb** [comment]
image

**Comment by Seelengrab** [comment]
This looks great! Could this be extended to multidimensional arrays and encouraging `axes`, if that's not already a thing that I just missed?

**Comment by fredrikekre** [comment]
I got a crash with:
```
ERROR: type Nothing has no field type
Stacktrace:
 [1] getproperty(x::Nothing, f::Symbol)
   @ Base ./Base.jl:42
 [2] check_incorrect_iter_spec(x::CSTParser.EXPR, body::CSTParser.EXPR, env::StaticLint.ExternalEnv)
   @ StaticLint ~/.julia/packages/StaticLint/ktd1a/src/linting/checks.jl:388
 [3] check_loop_iter(x::CSTParser.EXPR, env::StaticLint.ExternalEnv)
   @ StaticLint ~/.julia/packages/StaticLint/ktd1a/src/linting/checks.jl:357
```

**Comment by pfitzseb** [comment]
Should be fixed now.

**Review by fredrikekre — APPROVED** [review]

**Pull request merged**